File: iOS/Athlynk/Athlynk/Features/ExerciseDetailView.swift
//
//  ExerciseDetailView.swift
//  Stitch "Dettaglio Esercizio": full prescription for a single exercise —
//  media/video, target muscle, sets·reps·load·rpe·recovery·tempo, equipment,
//  and the coach's technique notes. Reuses ExerciseDTO from the workouts feed.
//

import SwiftUI

struct ExerciseDetailView: View {
    let exercise: ExerciseDTO
    @State private var appear = false
    @State private var stepsExpanded = false

    var body: some View {
        ZStack {
            VoltBackground(palette: [Palette.magenta, Palette.violet, Palette.cyan, Palette.magenta])
            ScrollView(showsIndicators: false) {
                VStack(alignment: .leading, spacing: 20) {
                    hero.revealUp(appear, index: 0)
                    titleBlock.revealUp(appear, index: 1)
                    statGrid.revealUp(appear, index: 2)
                    // Single execution text: instruction_steps only (description is
                    // deduplicated away). Collapsed to the first step until expanded.
                    if let steps = exercise.executionSteps, !steps.isEmpty {
                        instructionsBlock(steps).revealUp(appear, index: 4)
                    }
                    if !exercise.equipment.isEmpty { equipmentTag(exercise.equipment).revealUp(appear, index: 5) }
                    if let notes = exercise.techniqueNotes, !notes.isEmpty { coachNote(notes).revealUp(appear, index: 6) }
                    ExerciseTrendCard {
                        try await APIClient.shared.exerciseTrend(workoutExerciseId: exercise.id)
                    }
                    .revealUp(appear, index: 7)
                }
                .padding(.horizontal, 22).padding(.top, 12).padding(.bottom, AppLayout.tabBarClearance)
            }
        }
        .navigationTitle("")
        .navigationBarTitleDisplayMode(.inline)
        .toolbarBackground(.hidden, for: .navigationBar)
        .onAppear { appear = true }
    }

    private var hero: some View {
        ExerciseMediaHero(demoGifUrl: exercise.demoGifUrl, coverImageUrl: exercise.coverImageUrl,
                          videoUrl: exercise.videoUrl, accent: Palette.magenta)
    }

    private var titleBlock: some View {
        VStack(alignment: .leading, spacing: 6) {
            if let m = exercise.targetMuscleGroup, !m.isEmpty {
                Text("TARGET · \(m.uppercased())")
                    .font(Typo.mono(10, .bold)).tracking(2).foregroundStyle(Palette.magenta)
            }
            Text(exercise.name).font(Typo.poster(34)).foregroundStyle(Palette.textHi)
                .fixedSize(horizontal: false, vertical: true)
            Rectangle().fill(Palette.magenta).frame(height: 1).opacity(0.5)
        }
    }

    private var statGrid: some View {
        LazyVGrid(columns: [GridItem(.flexible()), GridItem(.flexible())], spacing: 12) {
            stat("square.stack.3d.up.fill", exercise.setCount.map { "\($0)" } ?? "—", "SERIE")
            stat("repeat", exercise.repRange ?? exercise.repCount.map { "\($0)" } ?? "—", "RIPETIZIONI")
            if let load = exercise.loadLabel { stat("scalemass.fill", load, "CARICO") }
            if let rec = exercise.recoverySeconds { stat("timer", "\(rec)s", "RECUPERO") }
            if let rpe = exercise.rpe { stat("flame.fill", "\(rpe)", "RPE") }
            if let rir = exercise.rir { stat("gauge.medium", "\(rir)", "RIR") }
            if let tempo = exercise.tempo, !tempo.isEmpty { stat("metronome.fill", tempo, "TEMPO") }
        }
    }

    private func stat(_ icon: String, _ value: String, _ label: String) -> some View {
        HStack(spacing: 12) {
            Image(systemName: icon).font(.system(size: 18, weight: .black)).foregroundStyle(Palette.magenta)
                .frame(width: 26)
            VStack(alignment: .leading, spacing: 1) {
                Text(value).font(Typo.mono(18, .black)).foregroundStyle(Palette.textHi)
                Text(label).font(Typo.mono(8, .bold)).tracking(1).foregroundStyle(Palette.textLow)
            }
            Spacer()
        }
        .padding(14).voltPanel(Palette.magenta.opacity(0.35))
    }

    private func equipmentTag(_ eq: [String]) -> some View {
        HStack(spacing: 8) {
            Image(systemName: "dumbbell.fill").font(.system(size: 13, weight: .bold))
            Text(eq.joined(separator: " · ").uppercased()).font(Typo.mono(11, .bold)).tracking(1)
        }
        .foregroundStyle(Palette.textMid)
        .padding(.horizontal, 12).padding(.vertical, 8)
        .background(Capsule().fill(Palette.void2))
    }

    /// Single execution text. Shows the first step; the rest reveal on tapping
    /// "vedi esecuzione" (collapsed by default).
    private func instructionsBlock(_ steps: [String]) -> some View {
        let visible = stepsExpanded ? Array(steps.enumerated()) : Array(steps.prefix(1).enumerated())
        return VStack(alignment: .leading, spacing: 12) {
            HStack(spacing: 8) {
                Image(systemName: "list.number").font(.system(size: 14, weight: .black))
                    .foregroundStyle(Palette.cyan)
                Text("ESECUZIONE").voltEyebrow()
            }
            VStack(alignment: .leading, spacing: 8) {
                ForEach(visible, id: \.offset) { i, step in
                    HStack(alignment: .top, spacing: 10) {
                        Text("\(i + 1)").font(Typo.mono(12, .black)).foregroundStyle(Palette.cyan)
                            .frame(width: 18, alignment: .leading)
                        Text(step).font(Typo.body(14)).foregroundStyle(Palette.textHi)
                            .lineLimit(stepsExpanded ? nil : 1)
                    }
                }
            }
            if steps.count > 1 {
                Button {
                    withAnimation(.easeInOut(duration: 0.2)) { stepsExpanded.toggle() }
                } label: {
                    HStack(spacing: 6) {
                        Text(stepsExpanded ? "Nascondi" : "Vedi esecuzione")
                            .font(Typo.mono(11, .bold)).tracking(1)
                        Image(systemName: stepsExpanded ? "chevron.up" : "chevron.down")
                            .font(.system(size: 10, weight: .black))
                    }
                    .foregroundStyle(Palette.cyan)
                }
            }
        }
        .padding(18).voltPanel(Palette.cyan.opacity(0.35))
    }

    private func coachNote(_ notes: String) -> some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack(spacing: 8) {
                Image(systemName: "quote.opening").font(.system(size: 14, weight: .black))
                    .foregroundStyle(Palette.cyan)
                Text("NOTA TECNICA").voltEyebrow()
            }
            Text(notes).font(Typo.body(15)).foregroundStyle(Palette.textHi).italic()
        }
        .padding(18).voltPanel(Palette.cyan.opacity(0.35))
    }
}

File: android/lib/features/athlete/workouts/exercise_detail_view.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:url_launcher/url_launcher.dart';

import '../../../core/api/athlete_api.dart';
import '../../../core/models/models.dart';
import '../../../core/providers.dart';
import '../../../design/components/exercise_trend_card.dart';
import '../../../design/components/misc.dart';
import '../../../design/components/panel.dart';
import '../../../design/components/scaffold.dart';
import '../../../design/theme.dart';

/// Full exercise prescription — port of iOS `ExerciseDetailView`: media hero,
/// stat grid, description, numbered instructions, equipment, coach note,
/// self-loading trend chart.
class ExerciseDetailView extends StatelessWidget {
  const ExerciseDetailView({super.key, required this.exercise});

  final ExerciseDto exercise;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.transparent,
      appBar: AppBar(),
      body: ExerciseDetailBody(exercise: exercise),
    );
  }
}

/// Reusable exercise content (no Scaffold) — used standalone and embedded in
/// the day pager, one exercise per page.
class ExerciseDetailBody extends ConsumerWidget {
  const ExerciseDetailBody({super.key, required this.exercise});

  final ExerciseDto exercise;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final ex = exercise;
    return ScreenScroll(
        topPadding: 0,
        spacing: Space.element,
        children: [
          ScreenHeader(
            eyebrow: ex.targetMuscleGroup ?? 'Esercizio',
            title: ex.name,
            titleSize: 28,
          ),
          ExerciseMediaHero(demoGif: ex.demoGif, coverImage: ex.coverImage),
          if ((ex.videoUrl ?? '').isNotEmpty)
            NavListRow(
              title: 'Guarda il video',
              icon: Icons.play_circle_outline_rounded,
              accent: Palette.magenta,
              onTap: () => launchUrl(Uri.parse(ex.videoUrl!),
                  mode: LaunchMode.externalApplication),
            ),
          _statGrid(ex),
          if ((ex.techniqueNotes ?? '').isNotEmpty)
            VoltPanel(
              tint: Palette.amber.withValues(alpha: 0.4),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  const Eyebrow('Nota del coach'),
                  const SizedBox(height: 6),
                  Text(ex.techniqueNotes!,
                      style: Typo.body(14, FontWeight.w500)),
                ],
              ),
            ),
          // Single execution text (instruction_steps only — description is
          // deduplicated away). Collapsed to the first step by default.
          if (ex.executionSteps.isNotEmpty)
            ExecutionStepsPanel(steps: ex.executionSteps),
          if (ex.equipment.isNotEmpty)
            Wrap(
              spacing: 8,
              runSpacing: 8,
              children: [
                for (final eq in ex.equipment)
                  Container(
                    padding: const EdgeInsets.symmetric(
                        horizontal: 12, vertical: 6),
                    decoration: BoxDecoration(
                      color: Palette.void1,
                      borderRadius: BorderRadius.circular(999),
                      border: Border.all(color: Palette.line),
                    ),
                    child: Text(eq,
                        style: Typo.mono(
                            10, FontWeight.w600, Palette.textMid)),
                  ),
              ],
            ),
          ExerciseTrendCard(
            loader: () =>
                ref.read(apiClientProvider).exerciseTrend(ex.id),
          ),
        ],
    );
  }

  Widget _statGrid(ExerciseDto ex) {
    final stats = <(String, String)>[
      ('Serie', ex.setCount?.toString() ?? '—'),
      ('Ripetizioni', ex.repRange ?? ex.repCount?.toString() ?? '—'),
      ('Carico', ex.loadLabel ?? '—'),
      ('Recupero',
          ex.recoverySeconds == null ? '—' : '${ex.recoverySeconds}s'),
      if (ex.rpe != null) ('RPE', '${ex.rpe}'),
      if (ex.rir != null) ('RIR', '${ex.rir}'),
      if ((ex.tempo ?? '').isNotEmpty) ('Tempo', ex.tempo!),
    ];
    return GridView.count(
      crossAxisCount: 3,
      shrinkWrap: true,
      physics: const NeverScrollableScrollPhysics(),
      mainAxisSpacing: 10,
      crossAxisSpacing: 10,
      childAspectRatio: 1.45,
      children: [
        for (final (label, value) in stats)
          Container(
            decoration: voltPanel(radius: 14),
            padding: const EdgeInsets.all(8),
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                FittedBox(
                  child: Text(value, style: Typo.poster(19)),
                ),
                const SizedBox(height: 2),
                Text(label.toUpperCase(),
                    style: Typo.mono(7.5, FontWeight.w700, Palette.textLow)
                        .copyWith(letterSpacing: 1)),
              ],
            ),
          ),
      ],
    );
  }
}

/// Execution steps shown one-line-collapsed with a "Vedi esecuzione" toggle.
class ExecutionStepsPanel extends StatefulWidget {
  const ExecutionStepsPanel({super.key, required this.steps});

  final List<String> steps;

  @override
  State<ExecutionStepsPanel> createState() => _ExecutionStepsPanelState();
}

class _ExecutionStepsPanelState extends State<ExecutionStepsPanel> {
  bool _expanded = false;

  @override
  Widget build(BuildContext context) {
    final steps = widget.steps;
    final visible = _expanded ? steps : steps.take(1).toList();
    return VoltPanel(
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          const Eyebrow('Esecuzione'),
          const SizedBox(height: 10),
          for (final (i, step) in visible.indexed)
            Padding(
              padding: const EdgeInsets.only(bottom: 10),
              child: Row(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Container(
                    width: 22,
                    height: 22,
                    decoration: BoxDecoration(
                      shape: BoxShape.circle,
                      color: Palette.magenta.withValues(alpha: 0.1),
                    ),
                    alignment: Alignment.center,
                    child: Text('${i + 1}',
                        style:
                            Typo.mono(10, FontWeight.w700, Palette.magenta)),
                  ),
                  const SizedBox(width: 10),
                  Expanded(
                    child: Text(step,
                        maxLines: _expanded ? null : 1,
                        overflow: _expanded ? null : TextOverflow.ellipsis,
                        style: Typo.body(13.5, FontWeight.w400)),
                  ),
                ],
              ),
            ),
          if (steps.length > 1)
            GestureDetector(
              onTap: () => setState(() => _expanded = !_expanded),
              child: Padding(
                padding: const EdgeInsets.only(top: 2),
                child: Row(
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    Text(_expanded ? 'Nascondi' : 'Vedi esecuzione',
                        style:
                            Typo.mono(11, FontWeight.w700, Palette.magenta)),
                    const SizedBox(width: 4),
                    Icon(
                        _expanded
                            ? Icons.keyboard_arrow_up_rounded
                            : Icons.keyboard_arrow_down_rounded,
                        size: 16,
                        color: Palette.magenta),
                  ],
                ),
              ),
            ),
        ],
      ),
    );
  }
}
